docs/pages/enroll-resources/database-access/auto-user-provisioning/postgres.mdx

```diff
@@ -41,6 +41,12 @@ Note that the RDS database must have IAM authentication enabled.
 
 Refer to the [AWS documentation](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/UsingWithRDS.IAMDBAuth.DBAccounts.html) to make sure you are using the `rds_iam` role correctly.
 for more information.
+
+If the admin user needs to grant the `rds_superuser` role to auto-provisioned
+users, the admin user must also be a `rds_superuser`:
+```sql
+GRANT rds_superuser TO "teleport-admin" WITH ADMIN OPTION;
+```
```

Comment:
@GavinFrazar i recall you mentioned that this is the case. could you confirm?

Reply:
The short answer is: you have to grant `rds_superuser` to "teleport-admin" in order for it to be able to grant `rds_superuser` to others, but you do not need the `WITH ADMIN OPTION`.

If you do not grant `rds_superuser` to "teleport-admin", then on v16+ RDS postgres you must `GRANT <role> TO "teleport-admin" WITH ADMIN OPTION`, and you probably want to do `WITH ADMIN OPTION, INHERIT FALSE, SET FALSE` technically.
On v15- RDS and OSS Postgres, "teleport-admin" merely needs `CREATEROLE` to grant non-superuser roles to others.

Note that RDS is not consistent with OSS postgres.

I tested to confirm this behavior on both v14 RDS postgres and v16 and compared it with OSS v14 and v16 postgres. The versions matter, because postgres changed what `CREATEROLE` attribute allows you to do starting with v16.

Overall, I recommend we tell people to just grant any roles to "teleport-admin" with the admin option, because it will always work. Otherwise, things may stop working if/when they upgrade to v16+ postgres.

The longer answer:

On OSS postgres v16+, you can only grant a role to another role in two cases:
1. you have the `SUPERUSER` attribute
2. OR you have been granted the role `WITH ADMIN OPTION`. Note that if you create a role, then such a grant is made automatically for you with `INHERIT FALSE, SET FALSE, ADMIN OPTION`.

ON OSS postgres v15-, you can grant a role to another role in both of the above cases, plus one more:
- you have `CREATEROLE` attribute, and the role you are granting does not have `SUPERUSER` attribute.

FWIW, on v15- OSS postgres, it doesn't automatically grant the `ADMIN OPTION` when you create a role, presumably because it wasn't necessary before they changed how `CREATEROLE` works.

postgres v16+ docs https://www.postgresql.org/docs/current/sql-grant.html

> The ADMIN option allows the member to in turn grant membership in the role to others, and revoke membership in the role as well. Without the admin option, ordinary users cannot do that. A role is not considered to hold WITH ADMIN OPTION on itself. Database superusers can grant or revoke membership in any role to anyone. This option defaults to FALSE.

postgres v15 and before docs https://www.postgresql.org/docs/15/sql-grant.html

> If WITH ADMIN OPTION is specified, the member can in turn grant membership in the role to others, and revoke membership in the role as well. Without the admin option, ordinary users cannot do that. A role is not considered to hold WITH ADMIN OPTION on itself. Database superusers can grant or revoke membership in any role to anyone. **Roles having CREATEROLE privilege can grant or revoke membership in any role that is not a superuser.**

According to AWS docs, `rds_superuser` does not have `SUPERUSER`, which I just confirmed on my v14 RDS instance as well: https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Appendix.PostgreSQL.CommonDBATasks.Roles.rds_superuser.html

```
postgres=> \duS+
...
 rds_superuser               | Cannot login                                               | 
 rdsadmin                    | Superuser, Create role, Create DB, Replication, Bypass RLS+| 
...
```

I can also see that the `rdsadmin` role, which is reserved for internal RDS usage only, has the SUPERUSER attribute, i.e it's actually a postgres superuser.

RDS does not follow OSS postgres and likely has been forked imo.
`rds_superuser` is like a superuser without actually being a real postgres superuser. It's weird. They do it to provide a managed postgres cluster that doesn't let you break things, I guess.

Reply:
Thanks for all the details and testing!!! Totally agreed the `rds_superuser` is a weird hack implemented by RDS.

We still need `createrole` to create the`teleport-auto-user` role even the admin user has the admin option to other roles.

I think i need to update this section to include this part:

> On OSS postgres v16+, you can only grant a role to another role in two cases:
> 1. you have the SUPERUSER attribute
> 2. OR you have been granted the role WITH ADMIN OPTION. Note that if you create a role, then such a grant is made automatically for you with INHERIT FALSE, SET FALSE, ADMIN OPTION.

Then for 15-, i will include that you must be `rds_superuser` (without the admin option) to grant `rds_superuser`